firmware/lm32/heartbeat.c

```diff
@@ -0,0 +1,71 @@
+
+#include <generated/csr.h>
+#include <generated/mem.h>
+#include <hw/flags.h>
+#include <system.h>
+#include <time.h>
+
+#include "heartbeat.h"
+#include "processor.h"
+#include "hdmi_in0.h"
+#include "hdmi_in1.h"
+#include "pattern.h"
+
+static int heartbeat_status = 0;
+
+void hb_status(int val)
+{
+	if(val==1) { 
+		heartbeat_status = 1; 
+	}
+	else {
+		heartbeat_status = 0;
+	}		
+}
+
+void hb_service(int source)
+{
+	static int last_event;
+	static int counter;		
+		if (heartbeat_status==1) {	
+		if(elapsed(&last_event, identifier_frequency_read()/5)) {
+			hb_fill(counter, source);
+			if(counter==5) 
+				counter = 6;	//BLUE in color_bar[6]
+			else 
+				counter = 5;	//RED in color_bar[5]
+		}
+	}
+}
+
+void hb_fill(int color_v, int source)
+{
+	int addr, i, j;
+	volatile unsigned int *framebuffer = (unsigned int *)(MAIN_RAM_BASE + pattern_framebuffer_base());
+
+#ifdef CSR_HDMI_IN0_BASE
```

Comment:
Modified this. This also fixed some irregular toggle in pattern source. And with some changes, also the hdmi_in sources.